overlord/ifacestate/helpers_test.go

```diff
@@ -333,3 +343,70 @@ func (s *helpersSuite) TestCheckIsSystemSnapPresentWithSnapd(c *C) {
 
 	c.Assert(ifacestate.CheckSystemSnapIsPresent(s.st), Equals, true)
 }
+
+// Check what happens with system-key when security profile regeneration fails.
+func (s *helpersSuite) TestSystemKeyAndFailingProfileRegeneration(c *C) {
+	dirs.SetRootDir(c.MkDir())
+	defer dirs.SetRootDir("")
+
+	// Create a fake security backend with failing Setup method and mock all
+	// security backends away so that we only use this special one. Note that
+	// the backend is given a non-empty name as the interface manager skips
+	// test backends with empty name for convenience.
+	backend := &ifacetest.TestSecurityBackend{
+		BackendName: "BROKEN",
+		SetupCallback: func(snapInfo *snap.Info, opts interfaces.ConfinementOptions, repo *interfaces.Repository) error {
+			return errors.New("cannot setup security profile")
+		},
+	}
+	restore := ifacestate.MockSecurityBackends([]interfaces.SecurityBackend{backend})
+	defer restore()
+
+	// Create a mock overlord, mainly to have state.
+	ovld := overlord.Mock()
+	st := ovld.State()
+
+	// Put a fake snap in the state, we need to setup security for at least one
+	// snap to give the fake security backend a chance to fail.
+	yamlText := `
+name: test-snapd-canary
+version: 1
+apps:
+  test-snapd-canary:
+    command: bin/canary
+`
+	si := &snap.SideInfo{Revision: snap.R(1), RealName: "test-snapd-canary"}
+	snapInfo := snaptest.MockSnap(c, yamlText, si)
+	st.Lock()
+	snapst := &snapstate.SnapState{
+		SnapType: string(snap.TypeApp),
+		Sequence: []*snap.SideInfo{si},
+		Active:   true,
+		Current:  snap.R(1),
+	}
+	snapstate.Set(st, snapInfo.InstanceName(), snapst)
+	st.Unlock()
+
+	// Pretend that security profiles are out of date and mock the
+	// function that writes the new system key with one always panics.
+	restore = ifacestate.MockProfilesNeedRegeneration(func() bool { return true })
```

Comment:
I don't think you need this (and the mocking) - the real profilesNeedRegenerationImpl() will DTRT when you write a bogus system-key into the file.

Reply:
Yes but they will pull in more stuff from the real system to know what the system-key should be. I mocked it for the sake of isolation.

Reply:
It is not the purpose of this test to call the real code that inspects the system to compute the system key.